Add ability to fit individually, globally, or both

jess-farmer opened the pull request:
**Description of work:**

Add ability to specify whether peaks should be individually fit, globally fit, or both, as mentioned in #59 

To do:
- [ ] Fix how invalid spectra are excluded from global fit - currently all are included which leads to poor energy estimate

**To test:**
Run the calibration scripts with the different `SharedParameterFitType` options: `'Individual'`, `'Shared'`, and `'Both'`

MialLewis commented on unpackaged/vesuvio_calibration/tests/system/test_system_analysis.py:
```diff
@@ -246,10 +264,12 @@ def _assert_E1_parameters_match_expected(self, params_table, rel_tolerance, fit_
         theta_errors = self._assert_theta_parameters_match_expected(params_table, rel_tol_theta)
-        L1_errors = self._assert_L1_parameters_match_expected(params_table, rel_tol_L1)
+        L1_errors = None
+        if fit_type != 'Shared':
```
Remind me, why can't we test L1 parameters if the fit is shared?

jess-farmer replied:
I've tried writing tests for the global L1 values (which are calculated using the global E1 values), but found that the values didn't match the expected L1 parameters - I was going to bring this up with Matthew when we next speak to him to check that the global L1 values are being calculated in the correct way

MialLewis replied:
Ok sounds good, I'll take another look at this PR when I get a chance, we're currently in our release sprint so I have a bit of a to-do list!

MialLewis commented on unpackaged/vesuvio_calibration/tests/unit/test_calibrate_vesuvio_fit.py:
```diff
@@ -660,39 +660,35 @@ def test_fit_peaks_individual(self, group_workspaces_mock):
-    @patch('calibration_scripts.calibrate_vesuvio_fit.EVSCalibrationFit._shared_parameter_fit')
     @patch('calibration_scripts.calibrate_vesuvio_fit.GroupWorkspaces')
-    def test_fit_peaks_shared(self, group_workspaces_mock, shared_parameter_fit_mock):
+    def test_fit_peaks_shared(self, group_workspaces_mock):
```
We test here if `_fit_shared_peak` is called correctly upon cal to `_fit_peaks`, can we add some tests to test the `_fit_shared_peak` function itself?

This should be easier to do once it has been refactored into several functions (each with tests ideally).

MialLewis reviewed (changes requested):
Thanks for the PR Jess. I've requested some changes largely related to refactoring of code into smaller functions.

Functions well.

jess-farmer commented:
Thanks for the review @MialLewis - I've now addressed most of your comments and have pushed the changes I've made so far. I still need to add more tests, and will aim to get this done by the end of the week.
